scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,328 @@
+# ranges is like [[a_x, b_x], [a_y, b_y]]
+def cub(f, ranges, rule, tol):
+    global_est, global_err = rule.estimate(f, ranges)
+
+    regions = [RegionInfo(global_est, global_err, ranges)]
+
+    # print(global_err)
+
+    while _max_norm(global_err) > tol:
+        region = heapq.heappop(regions)
+        print(_max_norm(global_err), global_err, region.ranges)
+        # print("Considering region", region)
+
+        est_k = region.integral_estimate
+        err_k = region.error_estimate
+        ranges_k = region.ranges
+
+        split_ranges = []
+
+        for rng in ranges_k:
+            split_ranges.append(split_range(rng))
+
+        # print("split ranges:", split_ranges)
+
+        # this hurts my eyes
+        # sub_ranges is now like
+        #     array([[[-1.,  0.],
+        #     [-1.,  0.]],
+
+        #    [[-1.,  0.],
+        #     [ 0.,  1.]],
+
+        #    [[ 0.,  1.],
+        #     [-1.,  0.]],
+
+        #    [[ 0.,  1.],
+        #     [ 0.,  1.]]])
+        all_sub_ranges = np.array(list(itertools.product(*split_ranges)))
+
+        # print("all sub ranges:", all_sub_ranges)
+
+        global_est -= est_k
+        global_err -= err_k
+
+        for sub_ranges in all_sub_ranges:
+            # print(sub_ranges)
+            est_sub, err_sub = rule.estimate(f, sub_ranges)
+
+            global_est += est_sub
+            global_err += err_sub
+
+            new_region = RegionInfo(est_sub, err_sub, sub_ranges)
+            # print("attempting to push", new_region, "onto heap")
+            # print("regions is currently", regions)
+
+            heapq.heappush(regions, new_region)
+
+    return global_est, global_err
+
+
+# Not sure this is really needed, was running into issues comparing very small floats
+# in the heap so made this to make it more explicit
+class RegionInfo:
+    def __init__(self, integral_estimate, error_estimate, ranges):
+        self.integral_estimate = integral_estimate
+        self.error_estimate = error_estimate
+        self.ranges = ranges
+
+    def key(self):
+        return -np.linalg.norm(abs(self.error_estimate))
+
+    def __le__(self, other):
+        # this is comparing -self.error_estimate against -other.error_estimate
+        return self.key() < other.key()
+
+    def __lt__(self, other):
+        return self.__le__(other)
+
+    def __repr__(self):
+        return f"""Estimate: {self.integral_estimate}
+Errors: {self.error_estimate} (max {-self.key()})
+Ranges: {self.ranges}"""
+
+    def __str__(self):
+        return self.__repr__()
+
+
+# Slight modification of https://stackoverflow.com/a/11146645
+# TODO: Is there a more canonical way of finding the Cartesian product of two numpy
+# arrays?
+def _cartesian_product(arrays):
+    la = len(arrays)
+    dtype = np.result_type(*arrays)
+    arr = np.empty([len(a) for a in arrays] + [la], dtype=dtype)
+    for i, a in enumerate(np.ix_(*arrays)):
+        arr[..., i] = a
+    return arr.reshape(-1, la)
+
```

Comment:
I think what we want here is a clever use of `np.meshgrid`. I think something like this will work:

```
# points is a list of your 1D arrays of points

out = np.stack(np.meshgrid(*points, indexing = 'ij'), axis = 0)
out = np.reshape(out, (ndim, -1))
```